contracts/samples/TokenPaymaster.sol

```diff
@@ -1,113 +1,213 @@
+/// The contract reverts the inner user transaction in that case but keeps the charge.
+/// The contract also allows honest clients to prepay tokens at a higher price to avoid getting reverted.
+/// It also allows updating price configuration and withdrawing tokens by the contract owner.
+/// The contract uses an Oracle to fetch the latest token prices.
+/// @dev Inherits from BasePaymaster.
+contract TokenPaymaster is BasePaymaster, UniswapHelper, OracleHelper {
+
+    struct TokenPaymasterConfig {
+        /// @notice The price markup percentage applied to the token price (1e6 = 100%)
+        uint256 priceMarkup;
+
+        /// @notice Exchange tokens to native currency if the EntryPoint balance of this Paymaster falls below this value
+        uint256 minEntryPointBalance;
+    }
 
-/**
- * A sample paymaster that defines itself as a token to pay for gas.
- * The paymaster IS the token to use, since a paymaster cannot use an external contract.
- * Also, the exchange rate has to be fixed, since it can't reference an external Uniswap or other exchange contract.
- * subclass should override "getTokenValueOfEth" to provide actual token exchange rate, settable by the owner.
- * Known Limitation: this paymaster is exploitable when put into a batch with multiple ops (of different accounts):
- * - while a single op can't exploit the paymaster (if postOp fails to withdraw the tokens, the user's op is reverted,
- *   and then we know we can withdraw the tokens), multiple ops with different senders (all using this paymaster)
- *   in a batch can withdraw funds from 2nd and further ops, forcing the paymaster itself to pay (from its deposit)
- * - Possible workarounds are either use a more complex paymaster scheme (e.g. the DepositPaymaster) or
- *   to whitelist the account and the called method ids.
- */
-contract TokenPaymaster is BasePaymaster, ERC20 {
-
-    //calculated cost of the postOp
-    uint256 constant public COST_OF_POST = 15000;
-
-    address public immutable theFactory;
-
-    constructor(address accountFactory, string memory _symbol, IEntryPoint _entryPoint) ERC20(_symbol, _symbol) BasePaymaster(_entryPoint) {
-        theFactory = accountFactory;
-        //make it non-empty
-        _mint(address(this), 1);
-
-        //owner is allowed to withdraw tokens from the paymaster's balance
-        _approve(address(this), msg.sender, type(uint).max);
+    event ConfigUpdated(TokenPaymasterConfig tokenPaymasterConfig);
+
+    event UserOperationSponsored(address indexed user, uint256 actualTokenCharge, uint256 actualGasCost, uint256 actualTokenPrice);
+
+    event PostOpReverted(address indexed user, uint256 preCharge);
+
+    event Received(address indexed sender, uint256 value);
+
+    /// @notice All 'price' variables are multiplied by this value to avoid rounding up
+    uint256 private constant PRICE_DENOMINATOR = 1e26;
```

Comment:
what is the rationale behind using 1e26?

Reply:
At this point it constants are actually pretty random and will be finalised after we make some gas usage optimisations. Do you have suggestions on what value should be used and why?